pkg/resources/server.go

```diff
@@ -0,0 +1,282 @@
+	sockPath := filepath.Join(types.SockDir, rs.endPoint)
+	lis, err := net.Listen("unix", sockPath)
+	if err != nil {
+		glog.Errorf("error starting %s device plugin endpoint: %v", resourceName, err)
+		return err
+	}
+
+	// Register all services
+	pluginapi.RegisterDevicePluginServer(rs.grpcServer, rs)
+
+	go rs.grpcServer.Serve(lis)
+	// Wait for server to start by launching a blocking connection
+	conn, err := grpc.Dial(sockPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithTimeout(5*time.Second),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+
+	if err != nil {
+		glog.Errorf("error. unable to establish test connection with %s gRPC server: %v", resourceName, err)
+		return err
+	}
+	glog.Infof("%s device plugin endpoint started serving", resourceName)
+	conn.Close()
+
+	rs.triggerUpdate()
+
+	// Register with Kubelet.
+	err = rs.register()
+	if err != nil {
+		// Stop server
+		rs.grpcServer.Stop()
+		glog.Fatal(err)
+		return err
+	}
+	return nil
+}
+
+func (rs *resourceServer) restart() error {
+	resourceName := rs.resourcPool.GetResourceName()
+	glog.Infof("restarting %s device plugin server...", resourceName)
+	if rs.grpcServer == nil {
+		return fmt.Errorf("grpc server instance not found for %s", resourceName)
+	}
+	rs.grpcServer.Stop()
+	rs.grpcServer = nil
+	// Send terminate signal to ListAndWatch()
+	rs.termSignal <- true
+
+	rs.grpcServer = grpc.NewServer() // new instance of a grpc server
+	return rs.Start()
+}
+
+func (rs *resourceServer) Stop() error {
+	resourceName := rs.resourcPool.GetResourceName()
+	glog.Infof("stopping %s device plugin server...", resourceName)
+	if rs.grpcServer == nil {
+		return nil
+	}
+	// Send terminate signal to ListAndWatch()
+	rs.termSignal <- true
+	rs.stopWatcher <- true
+
+	rs.grpcServer.Stop()
+	rs.grpcServer = nil
+
+	return rs.cleanUp()
+}
+
+func (rs *resourceServer) Watch() {
+	// Watch for socket file; if not present restart server
+	sockPath := filepath.Join(types.SockDir, rs.endPoint)
+	for {
+		select {
+		case stop := <-rs.stopWatcher:
+			if stop {
+				return
+			}
+		default:
+			_, err := os.Lstat(sockPath)
+			if err != nil {
+				// Socket file not found; restart server
+				glog.Warningf("server endpoint not found %s", rs.endPoint)
+				glog.Warningf("most likely Kubelet restarted")
+				if err := rs.restart(); err != nil {
+					glog.Fatalf("unable to restart server %v", err)
+				}
+			}
+
+		}
+		// Sleep for some intervals; TODO: investigate on suggested interval
+		time.Sleep(time.Second * time.Duration(5))
+	}
+}
+
+func (rs *resourceServer) cleanUp() error {
+	sockPath := filepath.Join(types.SockDir, rs.endPoint)
+	if err := os.Remove(sockPath); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+func (rs *resourceServer) triggerUpdate() {
+	rp := rs.resourcPool
+	if rs.checkIntervals > 0 {
```

Comment:
Is it possible to capture the notification of PF link state change and then triggers the update accordingly? With checkIntervals, there might be some short period when link state is changed, but kubelet doesn't get updated state and still allocated unhealthy devices to pods.

Reply:
We can have a file watcher instead that monitors the operstat? Lets start an issue on this later once this PR is merged

Reply:
yep, sth like a watcher or a block wait to detect notification immediately when state changes.

Reply:
Cool, please submit a PR with an implementation if you can, That would be great. Thanks